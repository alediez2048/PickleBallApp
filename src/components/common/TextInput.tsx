import React, { forwardRef } from "react";
import {
  TextInput as RNTextInput,
  Text,
  StyleSheet,
  TextInputProps as RNTextInputProps,
  TouchableOpacity,
  Platform,
} from "react-native";
import { useTheme } from "@/contexts/ThemeContext";
import { ThemedView } from "@/components/common/ThemedView";
import { ThemedText } from "@/components/common/ThemedText";

interface TextInputProps extends Omit<RNTextInputProps, "style"> {
  label?: string;
  error?: string;
  helperText?: string;
  startIcon?: React.ReactNode;
  endIcon?: React.ReactNode;
  onEndIconPress?: () => void;
  containerStyle?: any;
  inputStyle?: any;
}

const TextInputComponent = forwardRef<RNTextInput, TextInputProps>(
  (
    {
      label,
      error,
      helperText,
      startIcon,
      endIcon,
      onEndIconPress,
      containerStyle,
      inputStyle,
      editable = true,
      accessibilityLabel,
      accessibilityHint,
      ...props
    },
    ref
  ) => {
    const { colors } = useTheme();

    // Determine the help text to show (error takes precedence)
    const helpText = error || helperText;
    const isInvalid = !!error;

    // Combine accessibility hints
    const combinedAccessibilityHint = [accessibilityHint, helpText]
      .filter(Boolean)
      .join(". ");

    return (
      <ThemedView style={[styles.container, containerStyle]}>
        {label && (
          <ThemedText
            style={[
              styles.label,
              { color: isInvalid ? colors.primary : colors.text },
            ]}
            accessibilityRole="text"
          >
            {label}
          </ThemedText>
        )}

        <ThemedView
          style={[
            styles.inputContainer,
            { backgroundColor: colors.background },
          ]}
        >
          {startIcon && (
            <ThemedView style={styles.iconContainer} accessibilityRole="image">
              {startIcon}
            </ThemedView>
          )}

          <RNTextInput
            ref={ref}
            testID="text-input"
            style={[
              styles.input,
              {
                backgroundColor: editable ? colors.soft : colors.background,
                borderColor: isInvalid ? colors.primary : colors.border,
                color: colors.text,
              },
              startIcon && styles.inputWithStartIcon,
              endIcon && styles.inputWithEndIcon,
              !editable && styles.inputDisabled,
              inputStyle,
            ]}
            editable={editable}
            placeholderTextColor={colors.secondary}
            accessibilityLabel={accessibilityLabel || label}
            accessibilityHint={combinedAccessibilityHint || undefined}
            accessibilityState={{
              disabled: !editable,
            }}
            {...props}
          />

          {endIcon &&
            (onEndIconPress ? (
              <TouchableOpacity
                testID="end-icon-button"
                style={styles.iconContainer}
                onPress={onEndIconPress}
                accessibilityRole="button"
                accessibilityLabel="Toggle password visibility"
              >
                {endIcon}
              </TouchableOpacity>
            ) : (
              <ThemedView testID="end-icon" style={styles.iconContainer}>
                {endIcon}
              </ThemedView>
            ))}
        </ThemedView>

        {helpText && (
          <ThemedText
            style={[
              styles.helpText,
              { color: isInvalid ? colors.primary : colors.text },
            ]}
            accessibilityRole="text"
            accessibilityLiveRegion={isInvalid ? "polite" : "none"}
          >
            {helpText}
          </ThemedText>
        )}
      </ThemedView>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    marginBottom: 16,
  },
  label: {
    fontSize: 14,
    fontWeight: "500",
    marginBottom: 4,
  },
  inputContainer: {
    flexDirection: "row",
    alignItems: "center",
    position: "relative",
  },
  input: {
    flex: 1,
    height: 44,
    borderWidth: 1,
    borderRadius: 8,
    paddingHorizontal: 12,
    fontSize: 16,
    ...Platform.select({
      ios: {
        paddingVertical: 12,
      },
      android: {
        paddingVertical: 8,
      },
    }),
  },
  inputWithStartIcon: {
    paddingLeft: 44,
  },
  inputWithEndIcon: {
    paddingRight: 44,
  },
  inputDisabled: {
    opacity: 0.6,
  },
  iconContainer: {
    position: "absolute",
    height: "100%",
    width: 44,
    justifyContent: "center",
    alignItems: "center",
  },
  helpText: {
    fontSize: 12,
    marginTop: 4,
    marginLeft: 4,
  },
});

TextInputComponent.displayName = "TextInput";

export const TextInput = TextInputComponent;
