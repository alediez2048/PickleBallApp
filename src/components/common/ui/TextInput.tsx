import React, { forwardRef } from "react";
import {
  View,
  TextInput as RNTextInput,
  Text,
  StyleSheet,
  TextInputProps as RNTextInputProps,
  TouchableOpacity,
  Platform,
} from "react-native";
import { withMemo } from "@/components/hoc/withMemo";
import { useTheme } from "@/contexts/ThemeContext";

interface TextInputProps extends Omit<RNTextInputProps, "style"> {
  label?: string;
  error?: string;
  helperText?: string;
  startIcon?: React.ReactNode;
  endIcon?: React.ReactNode;
  onEndIconPress?: () => void;
  containerStyle?: any;
  inputStyle?: any;
}

const TextInputComponent = forwardRef<RNTextInput, TextInputProps>(
  (
    {
      label,
      error,
      helperText,
      startIcon,
      endIcon,
      onEndIconPress,
      containerStyle,
      inputStyle,
      editable = true,
      accessibilityLabel,
      accessibilityHint,
      ...props
    },
    ref
  ) => {
    const { colors } = useTheme();

    // Determine the help text to show (error takes precedence)
    const helpText = error || helperText;
    const isInvalid = !!error;

    // Combine accessibility hints
    const combinedAccessibilityHint = [accessibilityHint, helpText]
      .filter(Boolean)
      .join(". ");

    return (
      <View style={[styles.container, containerStyle]}>
        {label && (
          <Text
            style={[
              styles.label,
              { color: isInvalid ? colors.primary : colors.text },
            ]}
            accessibilityRole='text'
          >
            {label}
          </Text>
        )}

        <View
          style={[
            styles.inputContainer,
            { backgroundColor: colors.background },
          ]}
        >
          {startIcon && (
            <View style={styles.iconContainer} accessibilityRole='image'>
              {startIcon}
            </View>
          )}

          <RNTextInput
            ref={ref}
            testID='text-input'
            style={[
              styles.input,
              {
                backgroundColor: editable
                  ? colors.secondary
                  : colors.background,
                borderColor: isInvalid ? colors.primary : colors.tabIconDefault,
                color: colors.text,
              },
              startIcon && styles.inputWithStartIcon,
              endIcon && styles.inputWithEndIcon,
              !editable && styles.inputDisabled,
              inputStyle,
            ]}
            editable={editable}
            placeholderTextColor={colors.tabIconDefault}
            accessibilityLabel={accessibilityLabel || label}
            accessibilityHint={combinedAccessibilityHint || undefined}
            accessibilityState={{
              disabled: !editable,
            }}
            {...props}
          />

          {endIcon &&
            (onEndIconPress ? (
              <TouchableOpacity
                testID='end-icon-button'
                style={styles.iconContainer}
                onPress={onEndIconPress}
                accessibilityRole='button'
                accessibilityLabel='Toggle password visibility'
              >
                {endIcon}
              </TouchableOpacity>
            ) : (
              <View testID='end-icon' style={styles.iconContainer}>
                {endIcon}
              </View>
            ))}
        </View>

        {helpText && (
          <Text
            style={[
              styles.helpText,
              { color: isInvalid ? colors.primary : colors.text },
            ]}
            accessibilityRole='text'
            accessibilityLiveRegion={isInvalid ? "polite" : "none"}
          >
            {helpText}
          </Text>
        )}
      </View>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    marginBottom: 16,
  },
  label: {
    fontSize: 14,
    fontWeight: "500",
    marginBottom: 4,
  },
  inputContainer: {
    flexDirection: "row",
    alignItems: "center",
    position: "relative",
  },
  input: {
    flex: 1,
    height: 44,
    borderWidth: 1,
    borderRadius: 8,
    paddingHorizontal: 12,
    fontSize: 16,
    ...Platform.select({
      ios: {
        paddingVertical: 12,
      },
      android: {
        paddingVertical: 8,
      },
    }),
  },
  inputWithStartIcon: {
    paddingLeft: 44,
  },
  inputWithEndIcon: {
    paddingRight: 44,
  },
  inputDisabled: {
    opacity: 0.6,
  },
  iconContainer: {
    position: "absolute",
    height: "100%",
    width: 44,
    justifyContent: "center",
    alignItems: "center",
  },
  helpText: {
    fontSize: 12,
    marginTop: 4,
    marginLeft: 4,
  },
});

TextInputComponent.displayName = "TextInput";

export const TextInput = withMemo(TextInputComponent);
