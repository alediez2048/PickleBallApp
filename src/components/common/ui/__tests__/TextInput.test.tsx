import React from "react";
import renderer from "react-test-renderer";
import { TextInput } from "../TextInput";
import { View } from "react-native";

describe("<TextInput />", () => {
  it("renders correctly with basic props", () => {
    const tree = renderer
      .create(
        <TextInput
          label="Username"
          placeholder="Enter username"
          value=""
          onChangeText={() => {}}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with value", () => {
    const tree = renderer
      .create(
        <TextInput
          label="Username"
          placeholder="Enter username"
          value="johndoe"
          onChangeText={() => {}}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with error message", () => {
    const tree = renderer
      .create(
        <TextInput
          label="Username"
          placeholder="Enter username"
          value=""
          onChangeText={() => {}}
          error="This field is required"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with helper text", () => {
    const tree = renderer
      .create(
        <TextInput
          label="Password"
          placeholder="Enter password"
          value=""
          onChangeText={() => {}}
          helperText="Must be at least 8 characters"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly when disabled", () => {
    const tree = renderer
      .create(
        <TextInput
          label="Username"
          placeholder="Enter username"
          value="johndoe"
          onChangeText={() => {}}
          editable={false}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with start icon", () => {
    const tree = renderer
      .create(
        <TextInput
          label="Email"
          placeholder="Enter email"
          value=""
          onChangeText={() => {}}
          startIcon={<View testID="start-icon" />}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with end icon", () => {
    const tree = renderer
      .create(
        <TextInput
          label="Search"
          placeholder="Search..."
          value=""
          onChangeText={() => {}}
          endIcon={<View testID="end-icon" />}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with secure text entry", () => {
    const tree = renderer
      .create(
        <TextInput
          label="Password"
          placeholder="Enter password"
          value="password123"
          onChangeText={() => {}}
          secureTextEntry={true}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("forwards refs correctly", () => {
    const ref = React.createRef<any>();
    
    const tree = renderer
      .create(
        <TextInput
          ref={ref}
          label="Username"
          placeholder="Enter username"
          value=""
          onChangeText={() => {}}
        />
      )
      .toJSON();
      
    expect(tree).toMatchSnapshot();
    expect(ref.current).toBeDefined();
  });
});
