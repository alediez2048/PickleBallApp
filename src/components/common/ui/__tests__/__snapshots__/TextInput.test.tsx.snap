// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> forwards refs correctly 1`] = `null`;

exports[`<TextInput /> renders correctly when disabled 1`] = `null`;

exports[`<TextInput /> renders correctly with basic props 1`] = `null`;

exports[`<TextInput /> renders correctly with end icon 1`] = `null`;

exports[`<TextInput /> renders correctly with error message 1`] = `null`;

exports[`<TextInput /> renders correctly with helper text 1`] = `null`;

exports[`<TextInput /> renders correctly with secure text entry 1`] = `null`;

exports[`<TextInput /> renders correctly with start icon 1`] = `null`;

exports[`<TextInput /> renders correctly with value 1`] = `null`;
